### ISSUE TITLE
speeding up covariance matrix update

### ISSUE THREAD

**Issue by hardmaru** [CMA-ES/pycma#21]
Hi

Thanks for making your library available. I noticed that it has been updated to v2 very recently.

When I'm using the cma library as is, I notice that the covariance update is rather slow when the number of parameters exceeds a few thousand, and rather unusable when the number of parameters is > 10000.

After doing some literature search, it seems there are some previous works on speeding this up to reduce the compute complexity from O(N^3) to O(N^2).

http://dl.acm.org/citation.cfm?id=1144082

I'm wondering if there are any options in the existing library for using such method, or any low-rank approximation options available to speed up the covariance matrix update process so that it can be usable for large number of param size?

I have been trying to use CMA-ES along with other algorithms that are simpler, but scale quickly for a large number of params (for example, simple GA type algorithms, and also one of OpenAI's recent paper on ES).

Best Regards.

**Comment by nikohansen** [comment]
Hi,

The computational complexity of the code you are running is ``O(N^2)`` per iteration (bar bugs), because the part with cubic complexity is done only every ``Theta(N)`` iteration. This has always been the recommended default approach to implement CMA-ES. You could run a few quick experiments to verify this behaviour. 

The `cma` module indeed provides a (new) way to reduce the complexity to `O(N)` based on 

[Akimoto and Hansen (2016): Online model selection for restricted covariance matrix adaptation](https://scholar.google.com/scholar?cluster=809741855340971414)

This is likely to be the best choice in the `pycma` library when dimension exceeds 2000-or-so.

Invocation is not entirely straight forward, but there are is an example to start with [in `cma.test` (line 268ff)](https://github.com/CMA-ES/pycma/blob/master/cma/test.py#L264).

**Comment by hardmaru** [comment]
Hi Nikolaus,

Thanks for the response, the cma.test examples are really helpful.  Btw, away from the topic, I noticed that in the previous section of the code, in line 255, you had `restricted_gaussian_sampler.GaussVkDSampler`, but I think it should be `GaussVDSampler` since you are testing `VD-CMA`.

I notice that the VkD-CMA takes quite a long time to initialize. For example, a model with 10000 parameters can take close to 5 minutes to initialize with a beefy GCP virtual machine with ~200GB ram and fairly recent cpu cores.  Here is a quick profiling experiment I ran on my machine:

```
import time
import cma
from cma import restricted_gaussian_sampler

def make_es_profiler(num_param):
  start_time = time.time()
  es = cma.CMAEvolutionStrategy( num_param * [1], 1,
                               {
                                'CMA_active': False,
                                'AdaptSigma': None,
                                'CMA_sampler': restricted_gaussian_sampler.GaussVkDSampler,
                                'verbose': -9,
                               })
  end_time = time.time()
  return end_time - start_time

for num in [100, 500, 1000, 1500, 2000, 2500, 3000, 4000, 5000, 10000, 11274]:
  print('num param', num, 'time taken (s)', make_es_profiler(num))
```

The results:

```
num param 100 time taken (s) 0.13789987564086914
num param 500 time taken (s) 0.21612834930419922
num param 1000 time taken (s) 0.9570491313934326
num param 1500 time taken (s) 2.02294659614563
num param 2000 time taken (s) 3.6911516189575195
num param 2500 time taken (s) 6.703071594238281
num param 3000 time taken (s) 9.801333904266357
num param 4000 time taken (s) 24.549622535705566
num param 5000 time taken (s) 54.68238401412964
num param 10000 time taken (s) 249.79459977149963
num param 11274 time taken (s) 348.2705271244049
```

I guess the tradeoff in O(N) performance is to to use more memory in the optimizer? In my experiments, where the model has 11274 parameters, using the VkDSampler resulted in much slower performance, I think perhaps despite being O(N), the time constant is fairly large over the learner time constant in front of the O(N^2) default cma settings. The default cma still seems usable at 11274 params, although it is still slow compared to simpler optimizers, so I will still try to use it whenever I can.

I'm also using simpler optimizers, like the variants of PGPE that only adapt sigma to the diagonals to see if the covariance is generally worth it for RL-style tasks.

Thanks again for the help.

**Comment by nikohansen** [comment]
> in line 255, you had restricted_gaussian_sampler.GaussVkDSampler, but I think it should be GaussVDSampler since you are testing VD-CMA.

nice catch, will be corrected. 

> I guess the tradeoff in O(N) performance is to to use more memory in the optimizer?

It is not supposed to be like that, O(N) is in time and in memory (bar bugs). Though to guarantee this the entire time in the VkD version, the max k has to be set to a constant, e.g. 20. 

> In my experiments, where the model has 11274 parameters, using the VkDSampler resulted in much slower performance, I think perhaps despite being O(N), the time constant is fairly large over the learner time constant in front of the O(N^2) default cma settings.

My first hunch is that may rather be a bug than a feature...

**Comment by youheiakimoto** [comment]
Hi,

the problem of slow initialization is due to _updateBDfromSM method, defined in CMAEvolutionStrategy, costing O(N^3). I fixed the problem, by retuning a fake covariance matrix and an eigen decomposition of it. (pull request sent.) But this is absolutely not the right solution.

**Comment by nikohansen** [comment]
Fixed with merge 4449df2f7f195173849e691b66aafa6c532cf83d

**Comment by hardmaru** [comment]
Thanks. This will help for now (even though it is not the entirely right solution).

**Comment by nikohansen** [comment]
Test fixed with f2bb551229d1f82e6bcbd0bc3e539f80f9b7bb6b. There is still a problem, in that the VK-CMA-ES test [here](https://github.com/CMA-ES/pycma/blob/master/cma/test.py#L249) *sometimes* fails.
